Userland/Libraries/LibWasm/WASI/Wasi.cpp

```diff
@@ -697,6 +697,62 @@ ErrorOr<Result<Timestamp>> Implementation::impl$clock_time_get(Configuration&, C
     return Result<Timestamp> { static_cast<u64>(ts.tv_sec) * nanoseconds_in_second + static_cast<u64>(ts.tv_nsec) };
 }
 
+ErrorOr<Result<FileStat>> Implementation::impl$fd_filestat_get(Configuration&, FD fd)
+{
+    int resolved_fd = -1;
+
+    auto mapped_fd = map_fd(fd);
+    mapped_fd.visit(
+        [&](PreopenedDirectoryDescriptor descriptor) {
+            auto& entry = preopened_directories()[descriptor.value()];
+            resolved_fd = entry.opened_fd.value_or_lazy_evaluated([&] {
+                DeprecatedString path = entry.host_path.string();
+                return open(path.characters(), O_DIRECTORY, 0755);
+            });
+            entry.opened_fd = resolved_fd;
+        },
+        [&](u32 fd) {
+            resolved_fd = fd;
+        },
+        [](UnmappedDescriptor) {});
+
+    if (resolved_fd < 0)
+        return errno_value_from_errno(errno);
+
+    struct stat stat_buf;
+    if (fstat(resolved_fd, &stat_buf) < 0)
+        return errno_value_from_errno(errno);
+
+    constexpr auto file_type_of = [](struct stat const& buf) {
+        if (S_ISDIR(buf.st_mode))
+            return FileType::Directory;
+        if (S_ISCHR(buf.st_mode))
+            return FileType::CharacterDevice;
+        if (S_ISBLK(buf.st_mode))
+            return FileType::BlockDevice;
+        if (S_ISREG(buf.st_mode))
+            return FileType::RegularFile;
+        if (S_ISFIFO(buf.st_mode))
+            return FileType::Unknown; // no Pipe? :yakfused:
+        if (S_ISLNK(buf.st_mode))
+            return FileType::SymbolicLink;
+        if (S_ISSOCK(buf.st_mode))
+            return FileType::SocketDGram; // :shrug:
```

Comment:
Either these two comments for FIFO and sockets do not explain a lot (-> delete) or they should be `FIXME`s

Reply:
It's not something we can fix (WASI spec missing "proper" enum values), and it deserves a comment imo.
I should probably put _that_ in the comment :sweat_smile: